README.md: add a documentation on how to configure and run the container

jouvin opened the pull request:
This is an attempt to add a minimal documentation to help using this container, based on IJCLab experience.

Fixes #3

jouvin committed:
README.md: add a documentation on how to configure and run the container

giacomini commented:
Thanks for the PR. I report here a comment from the chat conversion, in case it helps in reviewing the text.

The interface of the script is given by 5 environment variables:

- FETCH_CRL_TIMEOUT_SECS (default 5 seconds)
- FORCE_TRUST_ANCHORS_UPDATE (default unset): if unset (or empty string) the script exits immediately and doesn't do anything
- TRUST_ANCHORS_TARGET (default unset): location where the egi trustanchors are mounted and where the container /etc/grid-security/certificates is rsync'ed
- CA_BUNDLE_TARGET (default unset): location where the system ca bundles are kept and where the container /etc/pki is rsync'ed. in particular a bundle (tls-ca-bundle-all.pem) contains the usual system bundle plus the egi trust-anchors
- CA_BUNDLE_SECRET_TARGET (default unset): a kubernetes secret including tls-ca-bundle-all.pem

The script can also have a command line argument, which is equivalent to specifying TRUST_ANCHORS_TARGET.

giacomini commented on README.md:
```diff
@@ -2,5 +2,73 @@
+
+volumes:
+    cabundle:
+        external: true
```
In our compose files as far as I know we have never used this `external` directive. We tend to use compose files to run test suites in Continuous Integration, where the volumes are created on the fly and then removed. Would you add a note for this typical use, saying that `external` is not needed?

jouvin replied:
`external: true` is required if you want to the volume to survive the container deletion. When using a CI (not sure what does it mean exactly here), it may not be necessary or may cause problems. But in a context where you run this container along the others in a standard Docker/Podman environment, if you don't do this, you may end up with an unvailable on other containers if you delete `egi-trustanchors`. I tried to clarify this.

giacomini replied:
Maybe I was not clear when I mentioned the use of a compose during a CI job. As you can see for example in the [VOMS testsuite](https://github.com/italiangrid/voms-testsuite/blob/master/compose/docker-compose.ci.yml), the compose includes many services, some of them using the EGI trustanchors volume, which is initialized by this container. In that situation the volume lives only for the duration of the compose run and that's fine.

jouvin replied:
So I think we agree! It is basically what I wrote: in a CI it may not make sense to use `external: true` but in a production service relying on Podman/Docker you really don't want your volume to be destroyed: you just want it to be updated. This this parameter is required.

giacomini commented on README.md:
````diff
@@ -2,5 +2,73 @@
+
+```
+# podman command can be used if using Podman instead of Docker
+0 3 * * * root (date --iso-8601=seconds --utc; (/usr/bin/docker restart egi-trustanchors)) >> /var/log/egi-trustanchors-update.log 2>&1
````
I'm not sure what restart does, but it may play badly with this [line](https://github.com/indigo-iam/egi-trust-anchors-container/blob/cee1a17e75306fc35ca4357c0a0df736af62ffed/update-trust-anchors.sh#L26), note the `>>`. Please check that the `tls-ca-bundle-all.pem` file doesn't keep increasing. On the other hand I wonder why there is an append there.

jouvin replied:
`restart` works if the container has already been started once (at startup for example) and is still present but exited... I have to double check because in Podman at least, after some time it doesn't restart anymore because of a problem with the network... I'll try to update this after I clarified the issue I was yesterday.

jouvin replied:
@giacomini about your concern regarding `tls-ca-bundle-all.pem` size continuously increasing, I don't understand why it would happen. `update_ca_trust` is not using the existing version of `tls-ca-bundle-all.pem` as an input but the source files that were used to build it. So you can run it as many times as you want without an impact on the size (as long as the source files are not modified).

giacomini replied:
The script uses `>>` instead of `>`, which would append to a possibly existing `tls-ca-bundle-all.pem`. I don't know the reason why it's written that way, but I just want to be sure that it doesn't cause trouble in case of a container restart. Personally, I would use `>` to recreate the bundle from scratch every time the script is run.

jouvin replied:
You're perfectly right! I didn't check carefully the code. I think we need to change this (as you, I think it is a non sense) but to avoid any breaking change, I'd propose to add an environment variable that will allow to chose between appending or replacing `tls-ca-bundle-all.pem` and keep the current behavior as the default. I'll try to open a PR implementing this tomorrow.

giacomini replied:
I wonder how the current behaviour (appending) may be useful, though.

jouvin replied:
I thought exactly the same after writing my comment. I may implement the opposite: change the default but keep a way to restore the old behavior in case there is a good reason for it that we missed. And once we validated there is none, just remove it.

giacomini commented on README.md:
```diff
@@ -2,5 +2,73 @@
+## Cron job
+
+The container exits after updating the trust anchors so it is necessary to add a cron job that periodically restarts the container
+(typically once a day). If the container has been created at startup, for example by executing the compose file, it is enough to
```
As observed by @balciiberk, once the CA bundle is updated, one needs to reload NGINX.

jouvin replied:
I know @balciiberk said it. I think it is for the `cabundle` part that is probably read once at Nginx startup. OTOH, I never restarted a grid service after a trustanchors update and I guess it is also true for the Nginx voms module (it doesn't cache the files in `/etc/grid-securiy/certificates` but read them at each request or as a mechanism/advisory lock to detect changes). I have to discuss this in more details with @balciiberk abouth the problem they faced. It may also be an impact of no using the `external: true` as Nginx may loose access to the previous version of the volume in some circumstances in this case...

If we agree that the potential issue is about cabundle changes not seen by Nginx, as for me, I tend to prefer not to restart Nginx each time `egi-trustanchors` is run (there is no mechanism to identify that the cabundle was updated) as failure to restart it would make the whole INDIGO IAM + VOMS AA unavailable. For me it is a too high price compared to the risk of not having the correct information on a CA (new CA, deprecated one, CA with a modified cert). Changes coming from the EGI side are quite rare in fact. It is not like the trustanchors that contains the CA CRL that changes every day that are not taken into account by Nginx itself but by its VOMS module (which for me checks the CRL at each request, to be confirmed).

giacomini replied:
It's true that the VOMS module reloads the whole /etc/grid-security/certificates at each request, because the VOMS library does that. I personally think it is a big waste of resources, but that's the way it is and I doubt I'll change it.

For NGINX I have verified that the bundle specified with `ssl_client_certificate` is (re)loaded only at configuration time and not on every request. Doing a reload of NGINX (and not a restart) is not particularly dangerous, because if the reload fails NGINX continues to work with the previous valid configuration.

jouvin replied:
Good point! I always forget about the `reload` option. I'll update the doc with it.

jouvin replied:
Documentation updated.

jouvin committed:
README.md: document environment variables, address @giacomini remarks

- In particular, document the required environment variables

jouvin commented:
Updaded to address your comments and document the environment variables. I am not sure what the failed check is about and if it has anything to do with my modifications...

jouvin committed:
README.md: add pull_policy to service configuration

- Ensure that the last image is downloaded from the registry every time
  the containter is recreated

jouvin committed:
README.md: further clarifications

jouvin commented:
@giacomini I think I addressed all your concerns/remarks. Please confirm and if you agree, I think it would help to merge this PR as an initial documentation that we can update when making other changes like #5...

giacomini commented on README.md:
```diff
@@ -2,5 +2,83 @@
 
-This is a [CENTOS 7 utility container](https://hub.docker.com/r/indigoiam/egi-trustanchors) that contains fetch-crl and other utilities
-to provide up-to-date trust anchors to relying applications.
+This is [container](https://hub.docker.com/r/indigoiam/egi-trustanchors) contains fetch-crl and other utilities
```
```suggestion
This [container](https://hub.docker.com/r/indigoiam/egi-trustanchors) contains fetch-crl and other utilities
```

giacomini commented on README.md:
```diff
@@ -2,5 +2,83 @@
+- `FETCH_CRL_TIMEOUT_SECS` (default 5 seconds): timeout of fetch-crl operations
+- `FORCE_TRUST_ANCHORS_UPDATE` (default unset): **needs to be defined for the update to be executed**. If unset (or empty string) the script
+exits immediately and doesn't do anything
+- `TRUST_ANCHORS_TARGET` (default unset): i**needs to be defined to properly update the trustanchors**. Location where the egi trustanchors
```
```suggestion
- `TRUST_ANCHORS_TARGET` (default unset): **needs to be defined to properly update the trustanchors**. Location where the egi trustanchors
```
Do you edit with `vi` :-)

jouvin replied:
Yes!

giacomini commented on README.md:
````diff
@@ -2,5 +2,83 @@
+```
+
+*Note: the parameters `external: true` in volumes is recommended if you use Docker or Podman to ensure that the volumes are not deleted once the container
+is removed. If you run the container in the ocntext of a CI, you may want to omit it so that the volume is created on the fly.*
````
```suggestion
is removed. If you run the container in the context of a CI, you may want to omit it so that the volume is created on the fly.*
```

giacomini commented:
There are three typos to fix, then we can merge it. I may want to add something else later on, for example the fact that you can pass also a command line argument.

Thanks again for the patch.

giacomini commented:
The check doesn't pass because there is a CI job that runs to rebuild the image, but it involves private credentials so it's not run (at least I think that's the reason, but we'll check better after the merge).

jouvin committed:
README: fix typos

jouvin commented:
@giacomini should be ok now!
